runner/src/internal.rs

```diff
@@ -51,15 +51,90 @@ impl Context {
     }
 }
 
+/// The outcome of an individual step of execution.
+#[derive(PartialEq, Eq)]
+pub enum StatusResultValue {
+    Ok,
+    Error,
+    Skipped,
+}
+
+pub struct SingleStatusResult {
+    value: StatusResultValue,
+    command: String,
+    logs: String,
+}
+
+pub struct MultipleStatusResult {
+    statuses: Vec<Status>,
+}
+
+/// The result of a step of execution which may itself be single or multiple.
+pub enum StatusResult {
+    Single(SingleStatusResult),
+    Multiple(MultipleStatusResult),
+}
+
+/// The result of executing a `Runnable` step.
+pub struct Status {
+    name: String,
+    status: StatusResult,
+}
+
+/// A runnable task, which may be implemented as a single or multiple steps.
 pub trait Runnable {
-    fn run(&self, context: &Context);
+    fn run(&self, opt: &Opt) -> Status;
 }
 
 /// Type alias for a boxed, dynamically dispatched `Runnable` instance.
 pub type R = Box<dyn Runnable>;
 
-pub fn run(runnable: R, opt: Opt) {
-    runnable.run(&Context::root(opt))
+/// Run the provided runnable, and print out a summary of the final status (including subtasks).
+pub fn run_all(runnable: R, opt: &Opt) {
+    let root = Context::root(&opt);
+    let status = runnable.run(&opt);
+    print_status(&root, &status);
+}
```

Comment:
Does `runner` exit with `1`, so GCB would stop with an error, if something is wrong ?

Reply:
Good point, it does not do that now, but I'll add it in a future PR (this is not run on GCB yet).